module/zfs/zfs_vnops.c

```diff
@@ -4922,6 +4923,736 @@ convoff(struct inode *ip, flock64_t *lckdat, int  whence, offset_t offset)
+
+		for (uint64_t i = start; i < start + size; i += blksz) {
+			zfs_fiemap_add_impl(t, i, blksz, 0, 0, 0,
+			    FIEMAP_EXTENT_UNWRITTEN | FIEMAP_EXTENT_DELALLOC |
+			    FIEMAP_EXTENT_UNKNOWN);
+		}
+	} else {
+		zfs_fiemap_add_impl(t, start, size, 0, 0, 0,
+		    FIEMAP_EXTENT_UNWRITTEN | FIEMAP_EXTENT_DELALLOC |
+		    FIEMAP_EXTENT_UNKNOWN | FIEMAP_EXTENT_MERGED);
+	}
+}
+
+/*
+ * The entire file is sparse and there are no level zero blocks with data.
+ * In this case pretend that hole block pointers exist to maintain consistency
+ * in the reported output.  Either add a single unwritten extent for the
+ * entire length of the file.  Or when no merging is requested add the
+ * correct number of hole block pointers.  Only the first extent tree should
+ * be populated since only holes are being added.
+ */
+static void
+zfs_fiemap_add_sparse(zfs_fiemap_t *fm)
+{
+	avl_tree_t *t = &fm->fm_extent_trees[0];
+	uint64_t blksz = fm->fm_block_size;
+	uint64_t size = P2ROUNDUP(fm->fm_file_size, blksz);
+
+	if (fm->fm_flags & FIEMAP_FLAG_NOMERGE) {
+		for (uint64_t i = 0; i < size; i += blksz) {
+			zfs_fiemap_add_impl(t, i, blksz, 0, 0, 0,
+			    FIEMAP_EXTENT_UNWRITTEN);
+		}
+	} else {
+		zfs_fiemap_add_impl(t, 0, size, 0, 0, 0,
+		    size == blksz ? FIEMAP_EXTENT_UNWRITTEN :
+		    FIEMAP_EXTENT_UNWRITTEN | FIEMAP_EXTENT_MERGED);
+	}
+}
+
+/*
+ * Walk the block pointers for the provided object and assemble a tree
+ * of extents which describe the logical to physical mapping.  Additionally
+ * include dirty buffers for the object which will be written but have
+ * not yet have had space allocated on disk.
+ */
+int
+zfs_fiemap_assemble(struct inode *ip, zfs_fiemap_t *fm)
+{
+	znode_t *zp = ITOZ(ip);
+	zfsvfs_t *zfsvfs = ZTOZSB(zp);
+	zbookmark_phys_t czb;
+	txg_handle_t th;
+	dnode_t *dn;
+	spa_t *spa;
+	uint64_t open_txg, syncing_txg, dirty_txg;
+	int error;
+
+	ZFS_ENTER(zfsvfs);
+	ZFS_VERIFY_ZP(zp);
+
+	error = dnode_hold(zfsvfs->z_os, zp->z_id, FTAG, &dn);
+	if (error) {
+		ZFS_EXIT(zfsvfs);
+		return (error);
+	}
+
+	spa = dmu_objset_spa(dn->dn_objset);
+
+	if (fm->fm_flags & FIEMAP_FLAG_SYNC)
+		txg_wait_synced(spa_get_dsl(spa), 0);
+
+	/*
+	 * Lock the entire file against changes while assembling the FIEMAP.
+	 * Then hold open the TXG while generating a map of all pending frees
+	 * and dirty blocks.  This isn't strictly necessary but it is a
+	 * convenient way to determine the range of TXGs to check.
+	 */
+	rl_t *rl = zfs_range_lock(&zp->z_range_lock, 0, UINT64_MAX, RL_READER);
+	open_txg = txg_hold_open(spa_get_dsl(spa), &th);
+	syncing_txg = dirty_txg = spa_syncing_txg(spa);
+
+	(void) dbuf_generate_dirty_maps(dn, fm->fm_dirty_tree,
+	    fm->fm_free_tree, &dirty_txg, open_txg);
+
+	/*
+	 * When the currently syncing TXG could not be checked, likely because
+	 * the dnode was already synced, we need to wait for the syncing TXG
+	 * to fully complete in order to avoid using stale block pointers.
```

Comment:
How stale can the BPs be?  The FIEMAP data is going to be stale the instant it leaves the kernel (if not before), even if you held an exclusive lock on the whole file for the duration of the call.  With that in mind, I don't think it is critical to take a heavy-weight approach to returning an atomic snapshot of the file block allocations.  What is most important is that extents are returned for regions of the file where there is data on disk or in memory, but the actual blocks are not as critical (e.g. if the file is being overwritten).